cms/private_media/managers.py

```diff
@@ -0,0 +1,124 @@
+        """
+        to_update = []
+        for obj in objects:
+            if obj.privacy != intended_privacy:
+                obj.privacy = intended_privacy
+                to_update.append(obj)
+        if not to_update:
+            return 0
+
+        to_update = self.bulk_set_file_permissions(to_update, intended_privacy, save_changes=False)
+
+        count = self.bulk_update(
+            to_update,  # type: ignore[arg-type]
+            fields=["_privacy", "privacy_last_changed", "file_permissions_last_set"],
+        )
+
+        if count and get_frontend_cache_configuration():
+            sites = Site.objects.all()
+            serve_url_batch = PurgeBatch()
+            file_url_batch = PurgeBatch()
+            for obj in to_update:
+                serve_url_batch.add_urls(obj.get_privacy_controlled_serve_urls(sites))
+                file_url_batch.add_urls(obj.get_privacy_controlled_file_urls())
+            serve_url_batch.purge()
+            file_url_batch.purge()
+        return count
+
+    def bulk_make_public(self, objects: Iterable["PrivateMediaMixin"]) -> int:
+        """Make an iterable of objects of this type 'public' as efficiently as
+        possible. Returns the number of objects that were actually updated in
+        the process.
+        """
+        return self.bulk_set_privacy(objects, Privacy.PUBLIC)
+
+    def bulk_make_private(self, objects: Iterable["PrivateMediaMixin"]) -> int:
+        """Make an iterable of objects of this type 'private' as efficiently as
+        possible. Returns the number of objects that were actually updated
+        in the process.
+        """
+        return self.bulk_set_privacy(objects, Privacy.PRIVATE)
+
+    def bulk_set_file_permissions(
+        self, objects: Iterable["PrivateMediaMixin"], intended_privacy: Privacy, *, save_changes: bool = False
+    ) -> list["PrivateMediaMixin"]:
+        """For an itrerable of objects of this type, set the file permissions for all
+        related files to reflect `intended_privacy`. Returns a list of the provided objects,
+        with their `file_permissions_last_set` datetime updated if all related files
+        were successfully updated.
+        """
+        files_by_object: dict[PrivateMediaMixin, list[FieldFile]] = defaultdict(list)
+        all_files = []
+        objects = list(objects)
+
+        for obj in objects:
+            for file in obj.get_privacy_controlled_files():
+                all_files.append(file)
+                files_by_object[obj].append(file)
+
+        results = bulk_set_file_permissions(all_files, intended_privacy)
+
+        now = timezone.now()
+        for obj in objects:
+            files = files_by_object[obj]
+            if all(results.get(file) for file in files):
+                obj.file_permissions_last_set = now
+
+        if save_changes:
+            self.bulk_update(objects, ["file_permissions_last_set"])  # type: ignore[arg-type]
+
+        return objects
```

Comment:
Because it doesn't always save changes to the objects. When being used by the other manager methods (which already call `bulk_update()` to save changes to the DB), saving is not necessary. Where used in other processes, saving is necessary, and the number of updated items holds no value.